### PR TITLE
Implement ICDF Methods for Truncated Distributions

### PR DESCRIPTION
**Pull request by TheSkyentist**
Implementation of generic Inverse Cumulative Distribution Function (ICDF) methods for Truncated Distributions. 

Sampling method for Truncated Distribution already calculates the ICDF, so the initial commit is refactoring the code into its own method. Still needs to be tested so starting this as a draft. 

Motivated by #1937 for use with Nested Sampling with `jaxns`.

### CODE REVIEW

**Comment by TheSkyentist** [comment]
Fixed an issue where the icdf functions would return non-NaN values for inputs outside of the range [0,1], although the values in the range are correct. Now the icdf function returns a NaN for inputs outside the range [0,1], which makes it consistent with other distributions as well.

File: numpyro/distributions/truncated.py [hunk 1]
```diff
@@ -138,7 +142,11 @@ def sample(self, key, sample_shape=()):
         finfo = jnp.finfo(dtype)
         minval = finfo.tiny
         u = random.uniform(key, shape=sample_shape + self.batch_shape, minval=minval)
-        return self.base_dist.icdf(u * self._cdf_at_high)
+        return self.icdf(u)
+
+    def icdf(self, q):
+        ppf = self.base_dist.icdf(q * self._cdf_at_high)
+        return jnp.where(q > 1, jnp.nan, ppf)
```
Comment (fehiepsi) [inline numpyro/distributions/truncated.py:149]: why do we need to cover q > 1?

File: numpyro/distributions/truncated.py [hunk 2]
```diff
@@ -67,11 +67,15 @@ def sample(self, key, sample_shape=()):
...
+
+    def icdf(self, q):
         loc = self.base_dist.loc
         sign = jnp.where(loc >= self.low, 1.0, -1.0)
-        return (1 - sign) * loc + sign * self.base_dist.icdf(
-            (1 - u) * self._tail_prob_at_low + u * self._tail_prob_at_high
+        ppf = (1 - sign) * loc + sign * self.base_dist.icdf(
+            (1 - q) * self._tail_prob_at_low + q * self._tail_prob_at_high
         )
+        return jnp.where(q < 0, jnp.nan, ppf)
```
Comment (fehiepsi) [inline numpyro/distributions/truncated.py:78]: why this where?

File: numpyro/distributions/truncated.py [hunk 3]
```diff
@@ -235,19 +243,22 @@ def sample(self, key, sample_shape=()):
...
-        #   A = 2 * loc - icdf[(1 - u) * cdf(2*loc-low)) + u * cdf(2*loc - high)]
+        #   A = 2 * loc - icdf[(1 - q) * cdf(2*loc-low)) + q * cdf(2*loc - high)]
         loc = self.base_dist.loc
         sign = jnp.where(loc >= self.low, 1.0, -1.0)
-        return (1 - sign) * loc + sign * self.base_dist.icdf(
-            clamp_probs((1 - u) * self._tail_prob_at_low + u * self._tail_prob_at_high)
+        ppf = (1 - sign) * loc + sign * self.base_dist.icdf(
+            clamp_probs((1 - q) * self._tail_prob_at_low + q * self._tail_prob_at_high)
         )
+        return jnp.where(jnp.logical_or(q < 0, q > 1), jnp.nan, ppf)
```
Comment (fehiepsi) [inline numpyro/distributions/truncated.py:261]: like above, why this where is needed?

**Review by fehiepsi — COMMENTED** [review]
Thanks @TheSkyentist Could you double check if our tests already cover the change?

File: numpyro/distributions/truncated.py (hunk 2 above)
Comment (TheSkyentist) [reply numpyro/distributions/truncated.py:78]: Please see my above comment. If this is not implemented, the icdf value will return a numeric, non NaN value outside the input range of [0,1]. The behavior of other NumPyro distributions is to return NaN outside this range. A non-NaN value outside of the input range can potentially cause bugs and/or confusion. Of course in principle one shouldn’t call the icdf with a value outside this input range, but I figured it would be best practice to ensure expected behavior.

File: numpyro/distributions/truncated.py (hunk 3 above)
Comment (TheSkyentist) [reply numpyro/distributions/truncated.py:261]: See above comment :-)

File: numpyro/distributions/truncated.py (hunk 2 above)
Comment (TheSkyentist) [reply numpyro/distributions/truncated.py:78]: I could also imagine doing the following: If we're worried about the performance hit to `sample` because of the added `jnp.where` call, then we could wrap the un-masked `icdf` code in a separate helper function and only mask out-of-range values for the `icdf` method. If it's preferred to keep the behaviour where numeric values are returned when out of bounds, then I'm happy to revert the previous commit :-)

**Comment by TheSkyentist** [comment]
I think the tests cover the change at least as far as I can see, but perhaps I'm not sure what you mean, could you clarify slightly? Thanks :-)

File: numpyro/distributions/truncated.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/distributions/truncated.py:78]: I think having the check is fine. Just curious, why not checking both q < 0 and q > 1?

Comment (TheSkyentist) [reply numpyro/distributions/truncated.py:78]: If it’s a left/right truncated distribution, we only need to check on the relevant side. Otherwise the base distribution icdf handles if it’s out of range.

Comment (fehiepsi) [reply numpyro/distributions/truncated.py:78]: Thanks for explaining!!

**Review by fehiepsi — APPROVED** [review]
